spec/fastsync/vdd/fastsyncL2.tla

```diff
@@ -0,0 +1,676 @@
+    LET rBlocks == 
+        [h \in Heights |-> IF h \in failedRequests THEN NilPeer ELSE bPool.receivedBlocks[h]] IN
+    LET bStore == 
+        [h \in Heights |-> IF h \in failedRequests THEN NilBlock ELSE bPool.blockStore[h]] IN  
+    
+    IF peer \in bPool.peerIds     
+    THEN [bPool EXCEPT 
+            !.peerIds = pIds,
+            !.peerHeights = pHeights,
+            !.pendingBlocks = pBlocks,
+            !.receivedBlocks = rBlocks,
+            !.blockStore = bStore
+          ]                              
+    ELSE bPool
+
+\* Remove faulty peers.
+\* Returns new block pool. 
+RemoveManyPeers(peers, bPool) ==
+    LET pIds == bPool.peerIds \ peers IN
+    LET pHeights == 
+        [p \in AllPeerIds |-> IF p \in peers THEN NilHeight ELSE bPool.peerHeights[p]] IN 
+    
+    LET failedRequests == 
+        {h \in Heights: /\ h >= bPool.height 
+                        /\ bPool.pendingBlocks[h] \in peers
+                        /\ bPool.receivedBlocks[h] \in peers} IN
+    LET pBlocks == 
+        [h \in Heights |-> IF h \in failedRequests THEN NilPeer ELSE bPool.pendingBlocks[h]] IN  
+    LET rBlocks == 
+        [h \in Heights |-> IF h \in failedRequests THEN NilPeer ELSE bPool.receivedBlocks[h]] IN
+    LET bStore == 
+        [h \in Heights |-> IF h \in failedRequests THEN NilBlock ELSE bPool.blockStore[h]] IN  
+    
+    IF pIds /= bPool.peerIds     
+    THEN [bPool EXCEPT 
+            !.peerIds = pIds,
+            !.peerHeights = pHeights,
+            !.pendingBlocks = pBlocks,
+            !.receivedBlocks = rBlocks,
+            !.blockStore = bStore
+          ]                              
+    ELSE bPool
+
+\* Add a peer.
+AddPeer(peer, bPool) == 
+    [bPool EXCEPT !.peerIds = bPool.peerIds \union {peer}]
+             
+     
+(* 
+Handle StatusResponse message.
+If valid status response, update peerHeights.
+If invalid (height is smaller than the current), then remove peer.
+Returns new block pool.
+*)                              
+HandleStatusResponse(msg, bPool) == 
+    LET peerHeight == bPool.peerHeights[msg.peerId] IN     
+    
+    IF /\ msg.peerId \in bPool.peerIds 
+       /\ msg.height >= peerHeight  
+    THEN    \* a correct response 
+        LET pHeights == [bPool.peerHeights EXCEPT ![msg.peerId] = msg.height] IN
+        [bPool EXCEPT !.peerHeights = pHeights]                   
+    ELSE RemovePeer(msg.peerId, bPool)   \* the peer has sent us message with smaller height or peer is not in our peer list
+        
+        
+(* 
+Handle BlockResponse message.
+If valid block response, update blockStore, pendingBlocks and receivedBlocks.
+If invalid (unsolicited response or malfored block), then remove peer.
+Returns new block pool.
+*)  
+HandleBlockResponse(msg, bPool) ==      
+    LET h == msg.block.height IN
+    
+    IF /\ msg.peerId \in bPool.peerIds
+       /\ bPool.blockStore[h] = NilBlock 
```

Comment:
If a peer sends a duplicate response (or the network duplicates the response), then the peer is going to be removed. Is it fine?

Reply:
Yep. This is the assumption we make about correct peers and network.